consensus/serenity/serenity.go

```diff
@@ -32,31 +29,50 @@ var (
 	// errInvalidNonce is returned if the nonce is non-zero.
 	errInvalidNonce = errors.New("invalid nonce")
 
-	// errInvalidMixDigest is returned if a block's mix digest is non-zero.
-	errInvalidMixDigest = errors.New("non-zero mix digest")
 	// errInvalidUncleHash is returned if a block contains an non-empty uncle list.
 	errInvalidUncleHash = errors.New("non empty uncle hash")
+
+	errOlderBlockTime = errors.New("timestamp older than parent")
 )
 
 // Serenity Consensus Engine for the Execution Layer.
-// Note: the work is mostly done on the Consensus Layer, so nothing much is to be added on this side.
-type Serenity struct{}
-
-// New creates a new instance of the Serenity Engine
-func New() *Serenity {
-	return &Serenity{}
+// Serenity is a consensus engine that combines the eth1 consensus and proof-of-stake
+// algorithm. The transition rule is described in the eth1/2 merge spec:
+// https://eips.ethereum.org/EIPS/eip-3675
+//
+// Note: After the Merge the work is mostly done on the Consensus Layer, so nothing much is to be added on this side.
+type Serenity struct {
+	eth1Engine consensus.Engine // Original consensus engine used in eth1, e.g. ethash or clique
+}
+
+// New creates a new instance of the Serenity Engine with the given embedded eth1 engine.
+func New(eth1Engine consensus.Engine) *Serenity {
+	if _, ok := eth1Engine.(*Serenity); ok {
+		panic("nested consensus engine")
+	}
+	return &Serenity{eth1Engine: eth1Engine}
 }
 
 // Author implements consensus.Engine, returning the header's coinbase as the
 // proof-of-stake verified author of the block.
 func (s *Serenity) Author(header *types.Header) (common.Address, error) {
+	if !IsPoSHeader(header) {
+		return s.eth1Engine.Author(header)
+	}
 	return header.Coinbase, nil
 }
 
 // VerifyHeader checks whether a header conforms to the consensus rules of the
 // stock Ethereum serenity engine.
 func (s *Serenity) VerifyHeader(chain consensus.ChainHeaderReader, header *types.Header, seal bool) error {
-	// Retrieve parent from the chain
+	reached, err := IsTTDReached(chain, header.ParentHash, header.Number.Uint64()-1)
```

Comment:
POSHeaders ancestor wont have total difficulty unless they are preceeded by the terminal block

Reply:
Let's then save total difficulty for all blocks, including PoS ones.